[wasm] reduce eval in production code

pavelsavara opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/61287

pavelsavara committed:
reduce eval in production code

pavelsavara commented:
cc @kg

pavelsavara committed:
- fix for JSExport nested types
- update benchmark to use JSExport/JSImport instead of reflection

pavelsavara commented:
Sorry guys, I realized that there is still usage in the benchmark.
That also forced me to fix JSExport from nested class.

pavelsavara commented on src/mono/sample/wasm/browser-bench/Program.cs:
```diff
@@ -72,6 +75,13 @@ public static void SetTasks(string taskNames)
         }
 
+        [JSExport]
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Probably not but I'm not confident about it yet. There are many places like this. Let's do the research and solve them all at once.

pavelsavara committed:
fix console benchmark

lewing commented:
https://radekdoulik.github.io/WasmPerformanceMeasurements/?startDate=2022-07-11T19%3A31%3A11.000Z&endDate=2022-07-11T21%3A06%3A41.000Z&tasks=%2CAppStart&flavors=2

kg commented:
> https://radekdoulik.github.io/WasmPerformanceMeasurements/?startDate=2022-07-11T19%3A31%3A11.000Z&endDate=2022-07-11T21%3A06%3A41.000Z&tasks=%2CAppStart&flavors=2

The time increase in this doesn't make any sense given the change in question, seems like variance?

lewing commented:
https://github.com/dotnet/runtime/compare/cae954add0f01c85de7fef8cc41c2a11d1c920e1...4a2ebf204801dc099d4e58dcdaf715dcecedcfda is the only commit in the diff and it changes the test mechanism

pavelsavara commented:
Yes, I had to rewrite the benchmark to stop using eval code path and so I introduced dependency on JSExport. The implementation at the time was triggering static constructors of the containing class during runtime init. That's now fixed in https://github.com/dotnet/runtime/pull/77293

radekdoulik commented:
> https://radekdoulik.github.io/WasmPerformanceMeasurements/?startDate=2022-07-11T19%3A31%3A11.000Z&endDate=2022-07-11T21%3A06%3A41.000Z&tasks=%2CAppStart&flavors=2

yup, it is https://github.com/dotnet/runtime/issues/75598 and Pavel fixed it in https://github.com/dotnet/runtime/pull/77293. the static constructors were doing a lot of work, which was not needed during startup
